pkg/cache/tas_flavor_snapshot.go

```diff
@@ -689,38 +689,38 @@ func findBestFitDomainIdx(domains []*domain, count int32) int {
 	return bestFitIdx
 }
 
-func (s *TASFlavorSnapshot) findLevelWithFitDomains(levelIdx int, required bool, count int32, unconstrained bool) (int, []*domain, string) {
+func (s *TASFlavorSnapshot) findLevelWithFitDomains(levelIdx int, required bool, podSetSize int32, unconstrained bool) (int, []*domain, string) {
 	domains := s.domainsPerLevel[levelIdx]
 	if len(domains) == 0 {
 		return 0, nil, fmt.Sprintf("no topology domains at level: %s", s.levelKeys[levelIdx])
 	}
 	levelDomains := slices.Collect(maps.Values(domains))
 	sortedDomain := s.sortedDomains(levelDomains, unconstrained)
 	topDomain := sortedDomain[0]
-	if useBestFitAlgorithm(unconstrained) && topDomain.state >= count {
+	if useBestFitAlgorithm(unconstrained) && topDomain.state >= podSetSize {
 		// optimize the potentially last domain
-		topDomain = sortedDomain[findBestFitDomainIdx(sortedDomain, count)]
+		topDomain = sortedDomain[findBestFitDomainIdx(sortedDomain, podSetSize)]
 	}
-	if topDomain.state < count {
+	if topDomain.state < podSetSize {
 		if required {
-			return 0, nil, s.notFitMessage(topDomain.state, count)
+			return 0, nil, s.notFitMessage(topDomain.state, podSetSize)
 		}
 		if levelIdx > 0 && !unconstrained {
-			return s.findLevelWithFitDomains(levelIdx-1, required, count, unconstrained)
+			return s.findLevelWithFitDomains(levelIdx-1, required, podSetSize, unconstrained)
 		}
 		results := []*domain{}
-		remainingCount := count
+		remainingCount := podSetSize
 		for idx := 0; remainingCount > 0 && idx < len(sortedDomain) && sortedDomain[idx].state > 0; idx++ {
 			offset := 0
 			if useBestFitAlgorithm(unconstrained) && sortedDomain[idx].state >= remainingCount {
 				// optimize the last domain
 				offset = findBestFitDomainIdx(sortedDomain[idx:], remainingCount)
 			}
 			results = append(results, sortedDomain[idx+offset])
-			remainingCount -= sortedDomain[idx].state
+			remainingCount -= sortedDomain[idx+offset].state
```

Comment:
The code should subtract the chosen domain state from `remainingCount`. Old code subtracted not the chosen one, but next in line. It is the same for all domains, apart from the last one in BestFit. The old code worked ok, because `remainingCount` is a local variable, and we only care it is "not greater than 0" at the end of the function, so even if it was lower than 0 it was fine.

I'm fixing it to account for the optimized last domain correctly.